File: src/esp32_dash/sensors/sensorHandler.cpp
#include "esp32_dash/Sensors/sensorHandler.h"

void SensorHandler::handleMessage(const Message& msg, CarState& state) {
    // --- DEBUG START ---
//    Serial.print("[I2C Incoming] Raw Payload: ");
//    Serial.println(msg.payload);
    // --- DEBUG END ---
    if (msg.type == TYPE_DATA) {
        parseMiataSensors(msg.payload, state);
    }
}

void SensorHandler::parseMiataSensors(const char* payload, CarState& state) {
    int lightsBit = 0;
    int beamBit = 0;
    int retractBit = 0;
    int rawRpm = 0;
    int waterTemp = 0;
    float oilPressure = 0.0f;

    // Expected format from Nano: "L:%d,B:%d,R:%d,T:%f"
    // L=Lights, B=HighBeam, R=Retractor, T=Tacho
    if (sscanf(payload, "L:%d,B:%d,U:%d,R:%d,W:%d,O:%.1f",
               &lightsBit,
               &beamBit,
               &retractBit,
               &rawRpm,
               &waterTemp,
               &oilPressure) >= 3) {

        // Map state
        state.lightsOn = (bool)lightsBit;
        state.beamOn = (bool)beamBit;
        state.lightsUp = (bool)retractBit;
        state.RPM = rawRpm;
        state.WaterTempC = waterTemp;
        state.OilPressure = oilPressure;
    }
}

File: rpi/src/handlers/sensor_handler.py
from src.bus.message import Message, MessageType
from src.handlers.base_handler import BaseHandler
from src.car_state import SharedState


class SensorHandler(BaseHandler):

    def handle_message(self, message: Message, state: SharedState) -> None:
        if message.type == MessageType.TYPE_DATA:
            self._parse(message.payload, state)

    def _parse(self, payload: str, state: SharedState) -> None:
        # Format: "lights,beam,up,rpm,water,oil*10"
        # Example: "1,0,1,1500,88,32"
        try:
            parts = payload.split(",")
            state.update(
                lights_on=bool(int(parts[0])),
                beam_on=bool(int(parts[1])),
                lights_up=bool(int(parts[2])),
                rpm=int(parts[3]),
                water_temp_c=int(parts[4]),
                oil_pressure=int(parts[5]) / 10.0,
            )
        except (IndexError, ValueError) as e:
            print(f"[SensorHandler] Failed to parse payload '{payload}': {e}")